Iceberg connector can't read columns after Spark changed float type to double type in ORC and Parquet

Avro format works fine. 

ORC
```
Error opening Iceberg split hdfs://hadoop-master:9000/user/hive/warehouse/test_orc/data/00000-1-3a532069-43a3-4495-9c12-b282d82e2c26-00001.orc (offset=3, length=261): Malformed ORC file. Cannot read SQL type 'double' from ORC stream '.col' of type FLOAT with attributes {iceberg.id=1, iceberg.required=true} [hdfs://hadoop-master:9000/user/hive/warehouse/test_orc/data/00000-1-3a532069-43a3-4495-9c12-b282d82e2c26-00001.orc]
io.trino.spi.TrinoException: Error opening Iceberg split hdfs://hadoop-master:9000/user/hive/warehouse/test_orc/data/00000-1-3a532069-43a3-4495-9c12-b282d82e2c26-00001.orc (offset=3, length=261): Malformed ORC file. Cannot read SQL type 'double' from ORC stream '.col' of type FLOAT with attributes {iceberg.id=1, iceberg.required=true} [hdfs://hadoop-master:9000/user/hive/warehouse/test_orc/data/00000-1-3a532069-43a3-4495-9c12-b282d82e2c26-00001.orc]
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.createOrcPageSource(IcebergPageSourceProvider.java:753)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.createDataPageSource(IcebergPageSourceProvider.java:535)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.createPageSource(IcebergPageSourceProvider.java:313)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:49)
	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:308)
	at io.trino.operator.Driver.processInternal(Driver.java:411)
	at io.trino.operator.Driver.lambda$process$10(Driver.java:314)
	at io.trino.operator.Driver.tryWithLock(Driver.java:706)
	at io.trino.operator.Driver.process(Driver.java:306)
	at io.trino.operator.Driver.processForDuration(Driver.java:277)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:752)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:164)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:519)
	at io.trino.$gen.Trino_dev____20230110_064454_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: io.trino.orc.OrcCorruptionException: Malformed ORC file. Cannot read SQL type 'double' from ORC stream '.col' of type FLOAT with attributes {iceberg.id=1, iceberg.required=true} [hdfs://hadoop-master:9000/user/hive/warehouse/test_orc/data/00000-1-3a532069-43a3-4495-9c12-b282d82e2c26-00001.orc]
	at io.trino.orc.reader.ReaderUtils.invalidStreamType(ReaderUtils.java:45)
	at io.trino.orc.reader.ReaderUtils.verifyStreamType(ReaderUtils.java:32)
	at io.trino.orc.reader.FloatColumnReader.<init>(FloatColumnReader.java:78)
	at io.trino.orc.reader.ColumnReaders.createColumnReader(ColumnReaders.java:78)
	at io.trino.orc.OrcRecordReader.createColumnReaders(OrcRecordReader.java:588)
	at io.trino.orc.OrcRecordReader.<init>(OrcRecordReader.java:248)
	at io.trino.orc.OrcReader.createRecordReader(OrcReader.java:308)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.createOrcPageSource(IcebergPageSourceProvider.java:707)
	... 17 more
```

Parquet
```
Failed to read Parquet file: hdfs://hadoop-master:9000/user/hive/warehouse/test_parquet/data/00000-2-86b138b2-8355-414f-8ef4-4b711a555f29-00001.parquet
io.trino.spi.TrinoException: Failed to read Parquet file: hdfs://hadoop-master:9000/user/hive/warehouse/test_parquet/data/00000-2-86b138b2-8355-414f-8ef4-4b711a555f29-00001.parquet
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.handleException(IcebergPageSourceProvider.java:1387)
	at io.trino.plugin.iceberg.IcebergPageSourceProvider.lambda$createParquetPageSource$30(IcebergPageSourceProvider.java:1073)
	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:75)
	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:407)
	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:386)
	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:293)
	at io.trino.spi.Page.getLoadedPage(Page.java:224)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:314)
	at io.trino.operator.Driver.processInternal(Driver.java:411)
	at io.trino.operator.Driver.lambda$process$10(Driver.java:314)
	at io.trino.operator.Driver.tryWithLock(Driver.java:706)
	at io.trino.operator.Driver.process(Driver.java:306)
	at io.trino.operator.Driver.processForDuration(Driver.java:277)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:752)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:164)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:519)
	at io.trino.$gen.Trino_dev____20230110_064454_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.UnsupportedOperationException: io.trino.spi.type.DoubleType
	at io.trino.spi.type.AbstractType.writeLong(AbstractType.java:91)
	at io.trino.parquet.reader.FloatColumnReader.readValue(FloatColumnReader.java:33)
	at io.trino.parquet.reader.PrimitiveColumnReader.lambda$readValues$0(PrimitiveColumnReader.java:165)
	at io.trino.parquet.reader.PrimitiveColumnReader.processValues(PrimitiveColumnReader.java:221)
	at io.trino.parquet.reader.PrimitiveColumnReader.readValues(PrimitiveColumnReader.java:163)
	at io.trino.parquet.reader.PrimitiveColumnReader.readPrimitive(PrimitiveColumnReader.java:152)
	at io.trino.parquet.reader.ParquetReader.readPrimitive(ParquetReader.java:444)
	at io.trino.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:529)
	at io.trino.parquet.reader.ParquetReader.readBlock(ParquetReader.java:512)
	at io.trino.parquet.reader.ParquetReader.lambda$nextPage$3(ParquetReader.java:274)
	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:72)
	... 17 more
```

Supported column type changes in Iceberg are mentioned in 
https://iceberg.apache.org/docs/latest/spark-ddl/#alter-table--alter-column. Other types are fine as fa as I confirmed.

--- a/lib/trino-orc/src/main/java/io/trino/orc/reader/FloatColumnReader.java
+++ b/lib/trino-orc/src/main/java/io/trino/orc/reader/FloatColumnReader.java
@@ -24,8 +24,8 @@ import io.trino.orc.stream.InputStreamSource;
 import io.trino.orc.stream.InputStreamSources;
 import io.trino.spi.block.Block;
 import io.trino.spi.block.IntArrayBlock;
+import io.trino.spi.block.LongArrayBlock;
 import io.trino.spi.block.RunLengthEncodedBlock;
-import io.trino.spi.type.RealType;
 import io.trino.spi.type.Type;
 import org.openjdk.jol.info.ClassLayout;
 
@@ -43,7 +43,10 @@ import static io.trino.orc.metadata.Stream.StreamKind.PRESENT;
 import static io.trino.orc.reader.ReaderUtils.minNonNullValueSize;
 import static io.trino.orc.reader.ReaderUtils.verifyStreamType;
 import static io.trino.orc.stream.MissingInputStreamSource.missingStreamSource;
+import static io.trino.spi.type.DoubleType.DOUBLE;
 import static io.trino.spi.type.RealType.REAL;
+import static java.lang.Double.doubleToRawLongBits;
+import static java.lang.Float.intBitsToFloat;
 import static java.lang.Math.toIntExact;
 import static java.util.Objects.requireNonNull;
 
@@ -51,6 +54,8 @@ public class FloatColumnReader
         implements ColumnReader
 {
     private static final int INSTANCE_SIZE = toIntExact(ClassLayout.parseClass(FloatColumnReader.class).instanceSize());
+
+    private final Type type;
 
     private final OrcColumn column;
 
@@ -74,8 +79,8 @@ public class FloatColumnReader
     public FloatColumnReader(Type type, OrcColumn column, LocalMemoryContext memoryContext)
             throws OrcCorruptionException
     {
-        requireNonNull(type, "type is null");
-        verifyStreamType(column, type, RealType.class::isInstance);
+        this.type = requireNonNull(type, "type is null");
+        verifyStreamType(column, type, t -> t == REAL || t == DOUBLE);
 
         this.column = requireNonNull(column, "column is null");
         this.memoryContext = requireNonNull(memoryContext, "memoryContext is null");
@@ -116,7 +121,7 @@ public class FloatColumnReader
                 throw new OrcCorruptionException(column.getOrcDataSourceId(), "Value is null but present stream is missing");
             }
             presentStream.skip(nextBatchSize);
-            block = RunLengthEncodedBlock.create(REAL, null, nextBatchSize);
+            block = RunLengthEncodedBlock.create(type, null, nextBatchSize);
         }
         else if (presentStream == null) {
             block = readNonNullBlock();
@@ -131,7 +136,7 @@ public class FloatColumnReader
                 block = readNullBlock(isNull, nextBatchSize - nullCount);
             }
             else {
-                block = RunLengthEncodedBlock.create(REAL, null, nextBatchSize);
+                block = RunLengthEncodedBlock.create(type, null, nextBatchSize);
             }
         }
 
@@ -147,7 +152,13 @@ public class FloatColumnReader
         verifyNotNull(dataStream);
         int[] values = new int[nextBatchSize];
         dataStream.next(values, nextBatchSize);
-        return new IntArrayBlock(nextBatchSize, Optional.empty(), values);
+        if (type == REAL) {
+            return new IntArrayBlock(nextBatchSize, Optional.empty(), values);
+        }
+        if (type == DOUBLE) {
+            return new LongArrayBlock(nextBatchSize, Optional.empty(), convertToLongArray(values));
+        }
+        throw new VerifyError("Unsupported type " + type);
     }
 
     private Block readNullBlock(boolean[] isNull, int nonNullCount)
@@ -163,8 +174,22 @@ public class FloatColumnReader
         dataStream.next(nonNullValueTemp, nonNullCount);
 
         int[] result = ReaderUtils.unpackIntNulls(nonNullValueTemp, isNull);
+        if (type == REAL) {
+            return new IntArrayBlock(isNull.length, Optional.of(isNull), result);
+        }
+        if (type == DOUBLE) {
+            return new LongArrayBlock(isNull.length, Optional.of(isNull), convertToLongArray(result));
+        }
+        throw new VerifyError("Unsupported type " + type);
+    }
 
-        return new IntArrayBlock(isNull.length, Optional.of(isNull), result);
+    private static long[] convertToLongArray(int[] intValues)
+    {
+        long[] values = new long[intValues.length];
+        for (int i = 0; i < intValues.length; i++) {
+            values[i] = doubleToRawLongBits(intBitsToFloat(intValues[i]));
+        }
+        return values;
     }
 
     private void openRowGroup()

--- a/lib/trino-parquet/src/main/java/io/trino/parquet/reader/FloatColumnReader.java
+++ b/lib/trino-parquet/src/main/java/io/trino/parquet/reader/FloatColumnReader.java
@@ -17,6 +17,8 @@ import io.trino.parquet.PrimitiveField;
 import io.trino.spi.block.BlockBuilder;
 import io.trino.spi.type.Type;
 
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.RealType.REAL;
 import static java.lang.Float.floatToRawIntBits;
 
 public class FloatColumnReader
@@ -30,6 +32,14 @@ public class FloatColumnReader
     @Override
     protected void readValue(BlockBuilder blockBuilder, Type type)
     {
-        type.writeLong(blockBuilder, floatToRawIntBits(valuesReader.readFloat()));
+        if (type == REAL) {
+            type.writeLong(blockBuilder, floatToRawIntBits(valuesReader.readFloat()));
+        }
+        else if (type == DOUBLE) {
+            type.writeDouble(blockBuilder, valuesReader.readFloat());
+        }
+        else {
+            throw new VerifyError("Unsupported type " + type);
+        }
     }
 }

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
@@ -2459,6 +2459,49 @@ public class TestIcebergSparkCompatibility
         onTrino().executeQuery(format("DROP TABLE %s", trinoTableName));
     }
 
+    @Test(groups = {ICEBERG, PROFILE_SPECIFIC_TESTS}, dataProvider = "testSparkAlterColumnType")
+    public void testSparkAlterColumnType(StorageFormat storageFormat, String sourceColumnType, String sourceValueLiteral, String newColumnType, Object newValue)
+    {
+        String baseTableName = "test_spark_alter_column_type_" + randomNameSuffix();
+        String trinoTableName = trinoTableName(baseTableName);
+        String sparkTableName = sparkTableName(baseTableName);
+
+        onSpark().executeQuery("CREATE TABLE " + sparkTableName + " TBLPROPERTIES ('write.format.default' = '" + storageFormat + "') " +
+                "AS SELECT CAST(" + sourceValueLiteral + " AS " + sourceColumnType + ") AS col");
+
+        onSpark().executeQuery("ALTER TABLE " + sparkTableName + " ALTER COLUMN col TYPE " + newColumnType);
+
+        assertEquals(getColumnType(baseTableName, "col"), newColumnType);
+        assertThat(onSpark().executeQuery("SELECT * FROM " + sparkTableName)).containsOnly(row(newValue));
+        assertThat(onTrino().executeQuery("SELECT * FROM " + trinoTableName)).containsOnly(row(newValue));
+
+        onSpark().executeQuery("DROP TABLE " + sparkTableName);
+    }
+
+    @DataProvider
+    public static Object[][] testSparkAlterColumnType()
+    {
+        Object[][] alterColumnTypeData = new Object[][] {
+                {"integer", "2147483647", "bigint", 2147483647L},
+                {"float", "10.3", "double", 10.3},
+                {"float", "'NaN'", "double", Double.NaN},
+                {"decimal(5,3)", "'12.345'", "decimal(10,3)", BigDecimal.valueOf(12.345)}
+        };
+
+        return Stream.of(StorageFormat.values())
+                .flatMap(storageFormat -> Arrays.stream(alterColumnTypeData).map(data -> new Object[] {storageFormat, data[0], data[1], data[2], data[3]}))
+                .toArray(Object[][]::new);
+    }
+
+    private String getColumnType(String tableName, String columnName)
+    {
+        return (String) onTrino().executeQuery("SELECT data_type FROM " + TRINO_CATALOG + ".information_schema.columns " +
+                        "WHERE table_schema = '" + TEST_SCHEMA_NAME + "' AND " +
+                        "table_name = '" + tableName + "' AND " +
+                        "column_name = '" + columnName + "'")
+                .getOnlyValue();
+    }
+
     private int calculateMetadataFilesForPartitionedTable(String tableName)
     {
         String dataFilePath = (String) onTrino().executeQuery(format("SELECT file_path FROM iceberg.default.\"%s$files\" limit 1", tableName)).getOnlyValue();

